runtime/storage/storage.go

```diff
@@ -0,0 +1,151 @@
+package storage
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/rilldata/rill/runtime/pkg/gcputil"
+	"gocloud.dev/blob"
+	"gocloud.dev/blob/gcsblob"
+	"gocloud.dev/gcp"
+)
+
+type Client struct {
+	dataDirPath  string
+	bucketConfig *gcsBucketConfig
+	prefixes     []string
+}
+
+func New(dataDir string, bucketCfg map[string]any) (*Client, error) {
+	c := &Client{
+		dataDirPath: dataDir,
+	}
+
+	if len(bucketCfg) != 0 {
+		gcsBucketConfig := &gcsBucketConfig{}
+		err := mapstructure.WeakDecode(bucketCfg, gcsBucketConfig)
+		if err != nil {
+			return nil, err
+		}
+		c.bucketConfig = gcsBucketConfig
+	}
+	return c, nil
+}
+
+func MustNew(dataDir string, bucketCfg map[string]any) *Client {
+	c, err := New(dataDir, bucketCfg)
+	if err != nil {
+		panic(err)
+	}
+	return c
+}
+
+func (c *Client) WithPrefix(prefix string) *Client {
+	newClient := &Client{
+		dataDirPath:  c.dataDirPath,
+		bucketConfig: c.bucketConfig,
+	}
+	newClient.prefixes = append(newClient.prefixes, c.prefixes...)
+	newClient.prefixes = append(newClient.prefixes, prefix)
+	return newClient
+}
+
+func (c *Client) DataDir(elem ...string) string {
+	paths := []string{c.dataDirPath}
+	if c.prefixes != nil {
+		paths = append(paths, c.prefixes...)
+	}
+	paths = append(paths, elem...)
+	return filepath.Join(paths...)
+}
+
+func (c *Client) TempDir(elem ...string) string {
+	paths := []string{c.dataDirPath}
+	if c.prefixes != nil {
+		paths = append(paths, c.prefixes...)
+	}
+	paths = append(paths, "tmp")
+	paths = append(paths, elem...)
+	return filepath.Join(paths...)
+}
+
+func (c *Client) OpenBucket(ctx context.Context, elem ...string) (*blob.Bucket, bool, error) {
+	if c.bucketConfig == nil {
+		return nil, false, nil
+	}
```

Comment:
To be on the safe side:

```suggestion
	if len(c.bucketConfig) == 0 {
		return nil, false, nil
	}
```

Reply:
Actually `bucketConfig` is a parsed struct not the input map.

Reply:
I see!